hbase-server/src/main/java/org/apache/hadoop/hbase/ccsmap/ChunkPool.java

```diff
@@ -0,0 +1,208 @@
+        }
+        chunk = allocatePooledChunk();
+      }
+    }
+    return chunk;
+  }
+
+  private AbstractHeapChunk allocateUnpooledChunk(int len) {
+    // An unpooled chunk must be on heap.
+    AbstractHeapChunk chunk = chunkAllocator.allocateUnpooledChunk(len);
+    unpooledChunkUsed.addAndGet(len);
+    unpooledChunks.put(chunk.getChunkID(), chunk);
+    return chunk;
+  }
+
+  /**
+   * Reclaim a chunk if it is reusable, remove it otherwise.
+   * @param chunk a chunk not in use
+   */
+  public void reclaimChunk(AbstractHeapChunk chunk) {
+    //not support putback duplicate.
+    if (chunk.getHeapMode() == this.heapMode && chunk.isPooledChunk()) {
+      chunk.getByteBuffer().clear();
+      if (!chunkQueue.offer(chunk)) {
+        // Should not happen, since the queue has no size limit.
+        throw new IllegalStateException("Can't reclaim chunk");
+      }
+    } else {
+      unpooledChunks.remove(chunk.getChunkID());
+      unpooledChunkUsed.addAndGet(-chunk.getLimit());
+    }
+  }
+
+  private AbstractHeapChunk allocatePooledChunk() {
+    AbstractHeapChunk chunk = chunkAllocator.allocatePooledChunk(chunkSize);
+    chunkArray[(int) chunk.getChunkID()] = chunk;
+    currentChunkCounter.incrementAndGet();
+    return chunk;
+  }
+
+  public AbstractHeapChunk getChunkByID(long chunkID) {
+    return chunkID < maxChunkCount ? chunkArray[(int) chunkID] : unpooledChunks.get(chunkID);
+  }
+
+  @VisibleForTesting
+  long getCurrentChunkCounter() {
+    return currentChunkCounter.get();
+  }
+
+  @VisibleForTesting
+  long getUnpooledChunkUsed() {
+    return unpooledChunkUsed.get();
+  }
+
+  @VisibleForTesting
+  Queue<AbstractHeapChunk> getChunkQueue() {
+    return chunkQueue;
+  }
+
+  @VisibleForTesting
+  AbstractHeapChunk[] getChunkArray() {
+    return chunkArray;
+  }
+
+  @VisibleForTesting
+  Map<Long, AbstractHeapChunk> getUnpooledChunksMap() {
+    return unpooledChunks;
+  }
+
+  @VisibleForTesting
+  int getMaxChunkCount() {
+    return maxChunkCount;
+  }
+
+  public static ChunkPool initialize(Configuration conf) {
+    if (globalInstance != null) {
+      return globalInstance;
+    }
+    synchronized (initLock) {
+      if (globalInstance == null) {
+        ChunkPoolParameters parameters = new ChunkPoolParameters(conf);
+        globalInstance = new ChunkPool(parameters);
+        LOG.info("CCSMapMemstore's chunkPool initialized with " + parameters);
+      }
+    }
+    return globalInstance;
+  }
+
+  static class ChunkPoolParameters {
+    long capacity;
+    int chunkSize;
+    int initialCount;
+    HeapMode heapMode;
+
+    ChunkPoolParameters(Configuration conf) {
+      capacity = conf.getLong(CCSMapUtils.CHUNK_CAPACITY_KEY, Long.MIN_VALUE);
+      chunkSize = conf.getInt(CCSMapUtils.CHUNK_SIZE_KEY, Integer.MIN_VALUE);
+      initialCount = conf.getInt(CCSMapUtils.INITIAL_CHUNK_COUNT_KEY, Integer.MIN_VALUE);
+      heapMode = conf.getBoolean(CCSMapUtils.USE_OFFHEAP, false) ?
```

Comment:
These parts are not very sensible by now. I plan to do a sanity check at the higher level (after we proceed there), so we can make sure when CCSMap is on, all needed configurations must be set. Something like this:
```
sanityCheck();
ChunkPool. initialize(conf);
```

Reply:
That's why the default value looks ugly.